rspace/src/main/scala/coop/rchain/rspace/history/TrieCache.scala

```diff
@@ -0,0 +1,181 @@
+package coop.rchain.rspace.history
+import coop.rchain.rspace.Blake2b256Hash
+
+import scala.collection.immutable.{Map => ImmutableMap}
+import scala.collection.immutable.Seq
+import scala.collection.mutable.Map
+
+private[rspace] trait CachedItem[TItem] {
+  def asOption: Option[TItem]
+}
+
+private[rspace] case class LoadedItem[TItem](value: TItem) extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = Some(value)
+}
+
+private[rspace] case class StoredItem[TItem](value: TItem) extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = Some(value)
+}
+
+private[rspace] case class AbsentItem[TItem]() extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = None
+}
+
+private[rspace] class TrieCache[T, K, V](
```

Comment:
don't you feel that this class is complex enough to merit some unit testing?

Reply:
It is already covered vert well by existing unit-tests - there just few uncovered methods left, like `getEmptyRoot`, `clear` and `toMap` which aren't used at all, and implemented just for consistency and possibility to reuse the `TrieCache` in another similar scenarios.

Reply:
Is it possible to add a test that would enforce the removal of weak references and check whether the store still behaves correctly?

Reply:
Sounds interesting, will try implement it.